Automatically use Open GPU drivers when desired

yeazelm opened the pull request:
**Description of changes:**
This adds the ability to select the open driver to the kmod packages in addition to the proprietary drivers at runtime. This should enable newer instance types that support the open GPU driver to use it automatically while keeping the older instance types on the proprietary driver. 

This now uses `ExecCondition` to call `ghostdog match-nvidia-driver` to check which devices are present and then either stop the link/load for the non-desired driver or continue if the requested driver matches the requested driver. 

It is best reviewed in commit order. `ghostdog` can be reviewed separately, the two kmod packages are mirrors of each other, and then the `os` package commit glues it all together.

### Additional notes

I added in an additional "copy only" configuration to copy the modules but not load them for `nvidia-drm` and `nvidia-peermem`. This is purely optional and we can choose to still not provide them, but since we are building them and they are present, it might be useful to some use cases to make them loadable at runtime. I haven't been able to find a way to exercise them though so we can remove that configuration file from the PR and they will remain in the staging directory until we decide to pull them in.

**Testing done:**
Tested on g5.2xlarge and g3.4xlarge both k8s 1.30 and 1.25 to confirm that the correct driver is loaded and the smoke tests to ensure the drivers still worked on each node. 

g3.4xlarge
```
bash-5.1# cat /proc/driver/nvidia/version
NVRM version: NVIDIA UNIX x86_64 Kernel Module  535.183.06  Wed Jun 26 06:46:07 UTC 2024
GCC version:  gcc version 11.3.0 (Buildroot 2022.11.1)
```

g4d.24xlarge
```
bash-5.1# cat /proc/driver/nvidia/version
NVRM version: NVIDIA UNIX Open Kernel Module for x86_64  535.183.06  Release Build  (dvs-builder@U16-I3-A14-4-1)  Wed Jun 26 06:59:29 UTC 2024
GCC version:  gcc version 11.3.0 (Buildroot 2022.11.1)
```

Example output from the gpu test:
```
[root@gpu-tests-77lfz /]# ./run.sh

=========================================
  Running sample UnifiedMemoryPerf
=========================================

GPU Device 0: "Ampere" with compute capability 8.6
....

[simpleVoteIntrinsics]
GPU Device 0: "Ampere" with compute capability 8.6

> GPU device has 80 Multi-Processors, SM 8.6 compute capabilities

[VOTE Kernel Test 1/3]
        Running <<Vote.Any>> kernel1 ...
        OK

[VOTE Kernel Test 2/3]
        Running <<Vote.All>> kernel2 ...
        OK

[VOTE Kernel Test 3/3]
        Running <<Vote.Any>> kernel3 ...
        OK
        Shutting down...

=========================================
  Running sample vectorAdd
=========================================

[Vector addition of 50000 elements]
Copy input data from the host memory to the CUDA device
CUDA kernel launch with 196 blocks of 256 threads
Copy output data from the CUDA device to the host memory
Test PASSED
Done

=========================================
  Running sample warpAggregatedAtomicsCG
=========================================

GPU Device 0: "Ampere" with compute capability 8.6

CPU max matches GPU max

Warp Aggregated Atomics PASSED
```


Output for the g4d.24xlarge:
```
bash-5.1# systemctl | grep kernel
  sys-kernel-config.mount                                                                                                                       loaded active mounted Kernel Configuration File System
  sys-kernel-debug.mount                                                                                                                        loaded active mounted Kernel Debug File System
  sys-kernel-tracing.mount                                                                                                                      loaded active mounted Kernel Trace File System
  var-lib-kernel\x2ddevel-.overlay-lower.mount                                                                                                  loaded active mounted Kernel Development Sources (Read-Only)
  x86_64\x2dbottlerocket\x2dlinux\x2dgnu-sys\x2droot-usr-src-kernels.mount                                                                      loaded active mounted Kernel Development Sources (Read-Write)
  copy-open-gpu-kernel-modules.service                                                                                                          loaded active exited  Link additional kernel modules
  load-open-gpu-kernel-modules.service                                                                                                          loaded active exited  Load additional kernel modules
  systemd-udevd-kernel.socket                                                                                                                   loaded active running udev Kernel Socket

bash-5.1# journalctl -u copy-open-gpu-kernel-modules
Sep 15 22:17:15 localhost systemd[1]: Starting Link additional kernel modules...
Sep 15 22:17:18 localhost driverdog[1489]: 22:17:18 [INFO] Copied nvidia.ko
Sep 15 22:17:19 ip-192-168-81-80.us-west-2.compute.internal driverdog[1489]: 22:17:19 [INFO] Copied nvidia-modeset.ko
Sep 15 22:17:25 ip-192-168-81-80.us-west-2.compute.internal driverdog[1489]: 22:17:25 [INFO] Copied nvidia-uvm.ko
Sep 15 22:17:25 ip-192-168-81-80.us-west-2.compute.internal driverdog[1742]: 22:17:25 [INFO] Copied nvidia-drm.ko
Sep 15 22:17:25 ip-192-168-81-80.us-west-2.compute.internal driverdog[1742]: 22:17:25 [INFO] Copied nvidia-peermem.ko
Sep 15 22:17:25 ip-192-168-81-80.us-west-2.compute.internal systemd[1]: Finished Link additional kernel modules.

bash-5.1# journalctl -u link-tesla-kernel-modules
Sep 15 22:17:15 localhost systemd[1]: Starting Link additional kernel modules...
Sep 15 22:17:15 localhost ghostdog[1415]: tesla is not preferred driver: open-gpu
Sep 15 22:17:15 localhost systemd[1]: link-tesla-kernel-modules.service: Skipped due to 'exec-condition'.
Sep 15 22:17:15 localhost systemd[1]: Condition check resulted in Link additional kernel modules being skipped.
Sep 15 22:17:26 ip-192-168-81-80.us-west-2.compute.internal systemd[1]: Starting Link additional kernel modules...
Sep 15 22:17:26 ip-192-168-81-80.us-west-2.compute.internal ghostdog[1786]: tesla is not preferred driver: open-gpu
Sep 15 22:17:26 ip-192-168-81-80.us-west-2.compute.internal systemd[1]: link-tesla-kernel-modules.service: Skipped due to 'exec-condition'.
Sep 15 22:17:26 ip-192-168-81-80.us-west-2.compute.internal systemd[1]: Condition check resulted in Link additional kernel modules being skipped.
Sep 15 22:17:27 ip-192-168-81-80.us-west-2.compute.internal systemd[1]: Starting Link additional kernel modules...
Sep 15 22:17:27 ip-192-168-81-80.us-west-2.compute.internal ghostdog[1819]: tesla is not preferred driver: open-gpu
Sep 15 22:17:27 ip-192-168-81-80.us-west-2.compute.internal systemd[1]: link-tesla-kernel-modules.service: Skipped due to 'exec-condition'.
Sep 15 22:17:27 ip-192-168-81-80.us-west-2.compute.internal systemd[1]: Condition check resulted in Link additional kernel modules being skipped.

bash-5.1# journalctl -u load-tesla-kernel-modules
Sep 15 22:17:15 localhost systemd[1]: Starting Load additional kernel modules...
Sep 15 22:17:15 localhost ghostdog[1490]: tesla is not preferred driver: open-gpu
Sep 15 22:17:15 localhost systemd[1]: load-tesla-kernel-modules.service: Skipped due to 'exec-condition'.
Sep 15 22:17:15 localhost systemd[1]: Condition check resulted in Load additional kernel modules being skipped.
Sep 15 22:17:26 ip-192-168-81-80.us-west-2.compute.internal systemd[1]: Starting Load additional kernel modules...
Sep 15 22:17:26 ip-192-168-81-80.us-west-2.compute.internal ghostdog[1790]: tesla is not preferred driver: open-gpu
Sep 15 22:17:26 ip-192-168-81-80.us-west-2.compute.internal systemd[1]: load-tesla-kernel-modules.service: Skipped due to 'exec-condition'.
Sep 15 22:17:26 ip-192-168-81-80.us-west-2.compute.internal systemd[1]: Condition check resulted in Load additional kernel modules being skipped.
Sep 15 22:17:27 ip-192-168-81-80.us-west-2.compute.internal systemd[1]: Starting Load additional kernel modules...
Sep 15 22:17:27 ip-192-168-81-80.us-west-2.compute.internal ghostdog[1836]: tesla is not preferred driver: open-gpu
Sep 15 22:17:27 ip-192-168-81-80.us-west-2.compute.internal systemd[1]: load-tesla-kernel-modules.service: Skipped due to 'exec-condition'.
Sep 15 22:17:27 ip-192-168-81-80.us-west-2.compute.internal systemd[1]: Condition check resulted in Load additional kernel modules being skipped.

bash-5.1# journalctl -u load-open-gpu-kernel-modules
Sep 15 22:17:25 ip-192-168-81-80.us-west-2.compute.internal systemd[1]: Starting Load additional kernel modules...
Sep 15 22:17:26 ip-192-168-81-80.us-west-2.compute.internal driverdog[1746]: 22:17:26 [INFO] Updated modules dependencies
Sep 15 22:17:26 ip-192-168-81-80.us-west-2.compute.internal driverdog[1746]: 22:17:26 [INFO] Loaded kernel modules
Sep 15 22:17:26 ip-192-168-81-80.us-west-2.compute.internal systemd[1]: Finished Load additional kernel modules.
```

Output for the g3.4xlarge:
```
bash-5.1# systemctl | grep kernel
  sys-kernel-config.mount                                                                                                                       loaded active mounted Kernel Configuration File System
  sys-kernel-debug.mount                                                                                                                        loaded active mounted Kernel Debug File System
  sys-kernel-tracing.mount                                                                                                                      loaded active mounted Kernel Trace File System
  var-lib-kernel\x2ddevel-.overlay-lower.mount                                                                                                  loaded active mounted Kernel Development Sources (Read-Only)
  x86_64\x2dbottlerocket\x2dlinux\x2dgnu-sys\x2droot-usr-src-kernels.mount                                                                      loaded active mounted Kernel Development Sources (Read-Write)
  link-tesla-kernel-modules.service                                                                                                             loaded active exited  Link additional kernel modules
  load-tesla-kernel-modules.service                                                                                                             loaded active exited  Load additional kernel modules
  systemd-udevd-kernel.socket                                                                                                                   loaded active running udev Kernel Socket

bash-5.1# journalctl -u link-tesla-kernel-modules
Sep 15 22:17:18 localhost systemd[1]: Starting Link additional kernel modules...
Sep 15 22:17:24 ip-192-168-24-94.us-west-2.compute.internal driverdog[2516]: 22:17:24 [INFO] Linked object 'nvidia.o'
Sep 15 22:17:24 ip-192-168-24-94.us-west-2.compute.internal driverdog[2516]: 22:17:24 [INFO] Stripped object 'nvidia.o'
Sep 15 22:17:24 ip-192-168-24-94.us-west-2.compute.internal driverdog[2516]: 22:17:24 [INFO] Linked object 'nvidia-modeset.o'
Sep 15 22:17:24 ip-192-168-24-94.us-west-2.compute.internal driverdog[2516]: 22:17:24 [INFO] Stripped object 'nvidia-modeset.o'
Sep 15 22:17:24 ip-192-168-24-94.us-west-2.compute.internal driverdog[2516]: 22:17:24 [INFO] Linked nvidia-uvm.ko
Sep 15 22:17:25 ip-192-168-24-94.us-west-2.compute.internal driverdog[2516]: 22:17:25 [INFO] Linked nvidia.ko
Sep 15 22:17:25 ip-192-168-24-94.us-west-2.compute.internal driverdog[2516]: 22:17:25 [INFO] Linked nvidia-modeset.ko
Sep 15 22:17:25 ip-192-168-24-94.us-west-2.compute.internal systemd[1]: Finished Link additional kernel modules.

bash-5.1# journalctl -u load-tesla-kernel-modules
Sep 15 22:17:25 ip-192-168-24-94.us-west-2.compute.internal systemd[1]: Starting Load additional kernel modules...
Sep 15 22:17:26 ip-192-168-24-94.us-west-2.compute.internal driverdog[2984]: 22:17:26 [INFO] Updated modules dependencies
Sep 15 22:17:26 ip-192-168-24-94.us-west-2.compute.internal driverdog[2984]: 22:17:26 [INFO] Loaded kernel modules
Sep 15 22:17:26 ip-192-168-24-94.us-west-2.compute.internal systemd[1]: Finished Load additional kernel modules.

bash-5.1# journalctl -u copy-open-gpu-kernel-modules
Sep 15 22:17:18 localhost systemd[1]: Starting Link additional kernel modules...
Sep 15 22:17:18 localhost ghostdog[2217]: open-gpu is not preferred driver: tesla
Sep 15 22:17:18 localhost systemd[1]: copy-open-gpu-kernel-modules.service: Skipped due to 'exec-condition'.
Sep 15 22:17:18 localhost systemd[1]: Condition check resulted in Link additional kernel modules being skipped.
Sep 15 22:17:27 ip-192-168-24-94.us-west-2.compute.internal systemd[1]: Starting Link additional kernel modules...
Sep 15 22:17:27 ip-192-168-24-94.us-west-2.compute.internal ghostdog[3029]: open-gpu is not preferred driver: tesla
Sep 15 22:17:27 ip-192-168-24-94.us-west-2.compute.internal systemd[1]: copy-open-gpu-kernel-modules.service: Skipped due to 'exec-condition'.
Sep 15 22:17:27 ip-192-168-24-94.us-west-2.compute.internal systemd[1]: Condition check resulted in Link additional kernel modules being skipped.
Sep 15 22:17:28 ip-192-168-24-94.us-west-2.compute.internal systemd[1]: Starting Link additional kernel modules...
Sep 15 22:17:28 ip-192-168-24-94.us-west-2.compute.internal ghostdog[3056]: open-gpu is not preferred driver: tesla
Sep 15 22:17:28 ip-192-168-24-94.us-west-2.compute.internal systemd[1]: copy-open-gpu-kernel-modules.service: Skipped due to 'exec-condition'.
Sep 15 22:17:28 ip-192-168-24-94.us-west-2.compute.internal systemd[1]: Condition check resulted in Link additional kernel modules being skipped.

bash-5.1# journalctl -u load-open-gpu-kernel-modules
Sep 15 22:17:18 localhost systemd[1]: Starting Load additional kernel modules...
Sep 15 22:17:18 localhost ghostdog[2517]: open-gpu is not preferred driver: tesla
Sep 15 22:17:18 localhost systemd[1]: load-open-gpu-kernel-modules.service: Skipped due to 'exec-condition'.
Sep 15 22:17:18 localhost systemd[1]: Condition check resulted in Load additional kernel modules being skipped.
Sep 15 22:17:27 ip-192-168-24-94.us-west-2.compute.internal systemd[1]: Starting Load additional kernel modules...
Sep 15 22:17:27 ip-192-168-24-94.us-west-2.compute.internal ghostdog[3034]: open-gpu is not preferred driver: tesla
Sep 15 22:17:27 ip-192-168-24-94.us-west-2.compute.internal systemd[1]: load-open-gpu-kernel-modules.service: Skipped due to 'exec-condition'.
Sep 15 22:17:27 ip-192-168-24-94.us-west-2.compute.internal systemd[1]: Condition check resulted in Load additional kernel modules being skipped.
Sep 15 22:17:28 ip-192-168-24-94.us-west-2.compute.internal systemd[1]: Starting Load additional kernel modules...
Sep 15 22:17:28 ip-192-168-24-94.us-west-2.compute.internal ghostdog[3097]: open-gpu is not preferred driver: tesla
Sep 15 22:17:28 ip-192-168-24-94.us-west-2.compute.internal systemd[1]: load-open-gpu-kernel-modules.service: Skipped due to 'exec-condition'.
Sep 15 22:17:28 ip-192-168-24-94.us-west-2.compute.internal systemd[1]: Condition check resulted in Load additional kernel modules being skipped.
```

**Terms of contribution:**

By submitting this pull request, I agree that this contribution is dual-licensed under the terms of both the Apache License, version 2.0, and the MIT license.

yeazelm commented on sources/ghostdog/src/main.rs:
```diff
@@ -125,6 +241,46 @@ fn parse_device_name(device_info: &[u8], path: String) -> Result<String> {
+/// Given a PCI ID, search the Open GPU Supported Devices File to determine if the Open GPU Driver should be used
+fn find_preferred_driver(pci_id: String) -> Result<String> {
+    // This corresponds to T4 devices which are Turing based and supported, but currently do not show up in the supported-devices file
+    if pci_id == "10DE:1EB8" {
```
I noticed that the g5g instances also don't match up on the supported list https://github.com/NVIDIA/open-gpu-kernel-modules?tab=readme-ov-file#compatible-gpus but are the Turing generation of devices that should support it. I'll do some more testing to see if the driver works on these instances too.

yeazelm replied:
I was able to get light confirmation from NVIDIA that these devices should use the proprietary driver so I've removed this if statement.

larvacea commented on packages/kmod-5.15-nvidia/40-nvidia-gpu-driver-select.rules:
```diff
@@ -0,0 +1,14 @@
+SUBSYSTEM=="pci", OPTIONS="log_level=debug"
+SUBSYSTEM=="pci", ATTR{vendor}=="0x10de", IMPORT{program}=="/usr/bin/ghostdog match-nvidia-driver %s{vendor}", TAG+="systemd"
+
+# If telsa, nothing more needed, driverdog will do the link/load
```
nit (spelling): should be "If tesla"

larvacea commented on sources/ghostdog/src/main.rs:
```diff
@@ -125,6 +241,46 @@ fn parse_device_name(device_info: &[u8], path: String) -> Result<String> {
+    match open_gpu_devices {
+        SupportedDevicesConfiguration::OpenGpu(device_list) => {
+            for supported_device in device_list.iter() {
+                if supported_device.device_id == format!("0x{}", input_device.device_id) {
```
nit: perhaps move the format! out of the inner loop? The loop does not iterate over terribly many instances, so not a big deal.

larvacea commented on sources/ghostdog/src/main.rs:
```diff
@@ -63,6 +141,44 @@ fn run() -> Result<()> {
+                let pci_id = env::var("PCI_ID").context(error::MissingPciIdEnvSnafu)?;
+                let driver_choice = find_preferred_driver(pci_id)?;
+                // After choosing the driver, drop a marker file provide this decision to others on the system (like systemd)
+                if !Path::new(NVIDIA_RUN_DIR).exists() {
```
Should we create the directory in systemd.tmpfiles? I have seen very similar directory creation fail during boot (possibly this can race), and tmpfiles is the canonical way to get the directory structure we want.

bcressey replied:
> Should we create the directory in systemd.tmpfiles?

tmpfiles runs rather late - after `local-fs.target` is reached - and `ghostdog` can be invoked before that.

yeazelm replied:
This is what I found in my testing, this executes so early in boot that you need to be ready to create these things yourself. They are called out in a tmpfile configuration but ghostdog beat it consistently. The same goes for when copying the modules, the invocation typically happens before so making the directory just in case ends up ensuring it is there.

larvacea commented on sources/ghostdog/src/main.rs:
Nit: the comment on line 2 is no longer true.

yeazelm replied:
Good call out, it also makes the name less applicable but I don't think changing the name is warranted.

bcressey commented on packages/kmod-5.15-nvidia/40-nvidia-gpu-driver-select.rules:
```diff
@@ -0,0 +1,14 @@
+# For any NVIDIA devices, ask ghostdog to select the right driver
+SUBSYSTEM=="pci", OPTIONS="log_level=debug"
```
```suggestion
```

yeazelm replied:
Missed this in my clean up of debugging lines!

bcressey commented on packages/kmod-5.15-nvidia/40-nvidia-gpu-driver-select.rules:
```diff
@@ -0,0 +1,14 @@
+# For any NVIDIA devices, ask ghostdog to select the right driver
+SUBSYSTEM=="pci", OPTIONS="log_level=debug"
```
Hoist preliminary checks to the top of the rules file so you don't have to keep checking the same things:
```suggestion
ACTION!="add", GOTO="nvidia_end"
SUBSYSTEM!="pci", GOTO="nvidia_end"
ATTR{vendor}!="0x10de", GOTO="nvidia_end"
```

yeazelm replied:
Nice! Will update.

bcressey commented on packages/kmod-5.15-nvidia/40-nvidia-gpu-driver-select.rules:
```diff
@@ -0,0 +1,14 @@
+SUBSYSTEM=="pci", ATTR{vendor}=="0x10de", ENV{BOTTLEROCKET_NVIDIA_DRIVER}=="nvidia-open-gpu", GOTO="nvidia_open_gpu"
+
+LABEL="nvidia_open_gpu"
+SUBSYSTEM=="pci", ATTR{vendor}=="0x10de", ENV{SYSTEMD_WANTS}="nvidia-open-gpu-modules.service", TAG+="systemd", GOTO="nvidia_end"
```
Can be simplified to:
```suggestion
ENV{BOTTLEROCKET_NVIDIA_DRIVER}=="nvidia-open-gpu", ENV{SYSTEMD_WANTS}="nvidia-open-gpu-modules.service", TAG+="systemd"
```

yeazelm replied:
I left this in place in case I needed to add things like GSP or other changes but yes, that is much simpler! I'll update to this knowing that we may need to add this back if GSP is needed.

bcressey commented on packages/kmod-5.15-nvidia/40-nvidia-gpu-driver-select.rules:
```diff
@@ -0,0 +1,14 @@
+# For any NVIDIA devices, ask ghostdog to select the right driver
+SUBSYSTEM=="pci", OPTIONS="log_level=debug"
+SUBSYSTEM=="pci", ATTR{vendor}=="0x10de", IMPORT{program}=="/usr/bin/ghostdog match-nvidia-driver %s{vendor}", TAG+="systemd"
```
Is the `TAG+="systemd"` needed here?

```suggestion
IMPORT{program}=="/usr/bin/ghostdog match-nvidia-driver %s{vendor}", TAG+="systemd"
```

yeazelm replied:
I added this when having difficulties getting systemd to notice my rules, I'll confirm if this can be removed since I believe the ENV check now works.

bcressey commented on packages/kmod-6.1-nvidia/nvidia-open-gpu-modules.service.in:
```diff
@@ -0,0 +1,22 @@
+# This ensures the directory is created before copying. The directory is created by tmpfiles but udev kicks this off early in boot
+ExecStart=/usr/bin/mkdir -p __NVIDIA_MODULES__
+# This copies the files from the factory directory which won't be found by modprobe to the proper driver directory which driverdog 
+# will expect. This prevents auto-loading of the drivers when teh proprietary driver is required.
```
```suggestion
# will expect. This prevents auto-loading of the drivers when the proprietary driver is required.
```

bcressey commented on packages/kmod-6.1-nvidia/kmod-6.1-nvidia.spec:
```diff
@@ -87,7 +102,25 @@ install -p -m 0644 %{S:2} .
+pushd NVIDIA-Linux-%{_cross_arch}-%{tesla_ver}/kernel-open
+
+# This recipe was based in the NVIDIA yum/dnf specs:
+# https://github.com/NVIDIA/yum-packaging-precompiled-kmod
```
nit: you can move this comment up to avoid repeating it

bcressey commented on packages/kmod-6.1-nvidia/kmod-6.1-nvidia.spec:
```diff
@@ -111,6 +144,12 @@ rm nvidia{,-modeset,-peermem}.o
+
+# Grab the list of supported devices
+pushd NVIDIA-Linux-%{_cross_arch}-%{tesla_ver}/supported-gpus
+cat supported-gpus.json | jq -r '.chips[] | select(.features[] | contains("kernelopen"))' | jq -s '{"open-gpu": .}' > open-gpu-supported-devices.json
```
nit: useless use of `cat`
```suggestion
jq -r '.chips[] | select(.features[] | contains("kernelopen"))' supported-gpus.json | jq -s '{"open-gpu": .}' > open-gpu-supported-devices.json
```

Wouldn't hurt to add a check to ensure that you can pull out a leaf value you expect to find, in case the structure changes slightly.

yeazelm replied:
Good call!

bcressey commented on packages/kmod-6.1-nvidia/kmod-6.1-nvidia.spec:
```diff
@@ -135,30 +185,45 @@ install -p -m 0644 %{S:203} %{buildroot}%{_cross_unitdir}
 install -d %{buildroot}%{_cross_datadir}/nvidia/tesla/module-objects.d
 install -d %{buildroot}%{_cross_factorydir}/nvidia/tesla
 
+# Open driver
```
Here I find the "open driver" and "proprietary driver" comments confusing - you're just creating some directories - so I'd remove them.

bcressey commented on packages/kmod-6.1-nvidia/kmod-6.1-nvidia.spec:
```diff
@@ -238,6 +324,7 @@ popd
 %files tesla-%{tesla_major}
 %license NVidiaEULAforAWS.pdf
 %license fabricmanager-linux-%{fm_arch}-%{tesla_ver}-archive/usr/share/doc/nvidia-fabricmanager/third-party-notices.txt
+# dirs
```
nit: it is obvious these are dirs, they have the `%dir` macro in front
```suggestion
```

bcressey commented on packages/kmod-6.1-nvidia/kmod-6.1-nvidia.spec:
```diff
@@ -362,6 +450,13 @@ popd
 
+# Open GPU drivers supported devices file
+%{_cross_datadir}/nvidia/open-gpu-supported-devices.json
+# depmod configuration files for prefering the right driver on load
```
style-wise, I would pick only one place for the comments:
* up where sources are defined
* in the body of the spec where you are putting the sources into place
* here in files

I definitely don't prefer comments in the `%files` section - we're in pure manifest / bookkeeping mode at this point - but repeated comments are the larger nit.

yeazelm replied:
Ok, fair enough. I wasn't sure with such a large and complex spec file if we wanted to be clear where things were going. I'm happy to drop the comments after the first intro of the files.

bcressey commented on packages/kmod-6.1-nvidia/nvidia-open-gpu-modules.service.in:
```diff
@@ -0,0 +1,22 @@
+[Unit]
+Description=Copy Open GPU drivers to needed modules directory
```
These descriptions are printed to the console, and should be somewhat terse or else they'll be truncated:
```suggestion
Description=Copy NVIDIA open GPU drivers
```

bcressey commented on packages/kmod-6.1-nvidia/nvidia-open-gpu-modules.service.in:
```diff
@@ -0,0 +1,22 @@
+[Unit]
+Description=Copy Open GPU drivers to needed modules directory
+Before=load-kernel-modules.service
+After=systemd-tmpfiles-setup.service
+DefaultDependencies=no
```
Why `DefaultDependencies=no`? This is usually done to make something happen very early, but because you depend on `systemd-tmpfiles-setup.service`, it won't run until after `local-fs.target` which is essentially what you end up with by using default dependencies.

For most cases of `DefaultDependencies=no` you also need to add `Conflicts=shutdown.target`. It's less important for oneshots but still good to do so that people don't forget for the cases where it does matter.

bcressey commented on packages/kmod-6.1-nvidia/nvidia-open-gpu-modules.service.in:
```diff
@@ -0,0 +1,22 @@
+
+[Service]
+Type=oneshot
+# This ensures the directory is created before copying. The directory is created by tmpfiles but udev kicks this off early in boot
+ExecStart=/usr/bin/mkdir -p __NVIDIA_MODULES__
```
I don't understand this comment. You won't run until after tmpfiles runs so why does it matter that udev is what activates the job?

yeazelm replied:
This is sort of a belt and suspenders approach since I've gained some distrust around trusting systemd ordering. That is not a good reason though so I should either trust systemd or pull the dependency out of the file and run the mkdir myself. I'll work on confirming I can trust tmpfiles will work but ultimately, I think this might move into driverdog anyway so this file might disappear entirely depending on how that works out.

yeazelm replied:
I've removed this from the approach and this should not need these `mkdir` calls.

bcressey commented on sources/driverdog/src/main.rs:
```diff
@@ -282,7 +288,17 @@ fn load_modules_sets(
 ) -> Result<()> {
     // Update the modules.dep before we attempt to load kernel modules
-    let args: Vec<String> = Vec::new();
+    // We need to confirm if there is a specific modules-set and if so, add a preference for that into depmod
+    let mut args: Vec<String> = Vec::new();
+    if modules_sets.contains_key("nvidia-tesla")
+        && Path::new("/usr/share/nvidia/tesla-depmod.conf").exists()
+    {
+        args.push("--config=/usr/share/nvidia/tesla-depmod.conf".to_string());
+    } else if modules_sets.contains_key("nvidia-open-gpu")
+        && Path::new("/usr/share/nvidia/open-gpu-depmod.conf").exists()
+    {
+        args.push("--config=/usr/share/nvidia/open-gpu-depmod.conf".to_string());
+    }
```
`driverdog` is a general program to link / load modules and this behavior is too specific to NVIDIA drivers. If you need a depmod.conf file then it should be in the TOML config file rather than hardcoded.

yeazelm replied:
I agree. Adding it to the configuration file is a much better approach.

yeazelm replied:
I removed this since I was able to provide the functionality we need without this additional logic.

bcressey commented on sources/driverdog/src/main.rs:
```diff
@@ -292,10 +308,11 @@ fn load_modules_sets(
-    // Load all the modules sets if no target module was given
-    for driver_config in modules_sets.values() {
-        load_modules(driver_config)?;
+    } else {
+        // Load all the modules sets if no target module was given
+        for driver_config in modules_sets.values() {
+            load_modules(driver_config)?;
+        }
```
This looks like a behavior change, since previously it would run unconditionally.

If the behavior needs to change, please split that into a separate commit with an explanation of the fix.

yeazelm replied:
I accidentally squashed the separate fix commit in my rebase. I do believe this was a bug that I found. Before this change, driverdog would still loop over all configuration files in `/etc/drivers` and try to load them, even if you were specifying a single modules-set via the arguments. I believe this is not the intended behavior by the shape of the CLI arguments but it isn't strictly documented. I think we should break this out separately and I can even PR this as its own change because it isn't strictly needed for this work.

bcressey commented on packages/kmod-6.1-nvidia/nvidia-open-gpu-load-config.toml.in:
```diff
@@ -0,0 +1,16 @@
+[nvidia-open-gpu]
+lib-modules-path = "kernel/drivers/extra/video/nvidia/open-gpu"
+objects-source = "__NVIDIA_MODULES__"
+load-only = true
```
Can load-only be inferred from the fact that all of the link objects lists are empty?

That seems less error-prone.

yeazelm replied:
I think other comments shifting work to driverdog make this config option less meaningful. I'll remove this and shift to a more flexible configuration approach that can link/copy/load based upon the configuration file.

bcressey commented on sources/ghostdog/src/main.rs:
```diff
@@ -63,6 +141,44 @@ fn run() -> Result<()> {
             emit_device_name(&device_name);
         }
+        SubCommand::MatchNvidiaDriver(vendor) => {
+            // Currently, we only want to run this for NVIDIA devices
+            if vendor.vendor == NVIDIA_VENDOR_ID {
```
This seems redundant with the udev rule, which only invokes `ghostdog match-nvidia-driver` if the vendor ID matches.

yeazelm replied:
Because it will try to read files only provided by the nvidia kmod packages, I was trying to gate against proceeding to do something I know would not work on other variants. It is redundant with the udev rule, it was more around when *not* invoked by the udev rule which should be zero.

bcressey commented on sources/ghostdog/src/main.rs:
```diff
@@ -63,6 +141,44 @@ fn run() -> Result<()> {
+                let driver_choice = find_preferred_driver(pci_id)?;
+                // After choosing the driver, drop a marker file provide this decision to others on the system (like systemd)
+                if !Path::new(NVIDIA_RUN_DIR).exists() {
+                    fs::create_dir_all(NVIDIA_RUN_DIR).context(error::CreateDirSnafu {
+                        path: NVIDIA_RUN_DIR,
+                    })?;
+                }
+                let marker_path: PathBuf = Path::new(NVIDIA_RUN_DIR).join(driver_choice.clone());
+                fs::write(marker_path.clone(), "").context(error::WriteMarkerFileSnafu {
+                    path: marker_path.clone(),
+                })?;
```
I didn't understand at first why this marker is necessary - `nvidia-open-gpu-modules.service` should only be active if this udev rule adds it.

I've left more comments on that service, but I think it'd be better to avoid doing this and shift more work into `driverdog`.

yeazelm replied:
I think this is fair, this was mostly a way for systemd unit files to have Path conditions that could confirm this choice. I agree that moving some of this into driverdog is the right call.

yeazelm replied:
Moved away from this approach, `ghostdog` only worries about checking the driver and exits with a code to represent that.

bcressey commented on packages/kmod-6.1-nvidia/nvidia-open-gpu-modules.service.in:
```diff
@@ -0,0 +1,22 @@
+RemainAfterExit=true
+StandardError=journal+console
+
+[Install]
+WantedBy=load-kernel-modules.service
```
This is very racy - `systemd` will always queue the job immediately, and if it happens to start before `ghostdog` finishes creating the link, then the condition check won't pass and it will fail to start.

It looks like you're working around that by pushing the job after `systemd-tmpfiles-setup.service`, but that's a hack.

Zooming out a bit, it seems like you've created two synchronization problems for yourself:
* the need for the correct config file to be in place before either `driverdog link` or `driverdog load` can run
* the need to copy the .ko files into place before `driverdog load` runs, in the open GPU case

You can eliminate the second coordination problem here by making it a `driverdog` config problem instead - having a `copy-source` field instead of `link-objects` to tell `driverdog` where to find the `.ko`.

I know I pushed back on more `driverdog` functionality here but I can be wrong sometimes. It would let you delete this unit (and its sync problem) and focus only on the "correct config for driverdog" problem.

yeazelm replied:
I've resolved this by moving away from `udev` entirely. We can determine at the time of the unit file running what decision to make. It costs _very little_ to do the `ghostdog match-nvidia-driver` for these invocations so keeping track of these decisions somewhere else don't look to be worth it. Testing on a g3.4xlarge which should have to crawl the entire file and not make a match (worst case runtime) is still insignificant compared to the time it takes to link/load the modules:

```
bash-5.1# time ghostdog match-nvidia-driver tesla

real    0m0.005s
user    0m0.002s
sys     0m0.003s
bash-5.1# time ghostdog match-nvidia-driver open-gpu
open-gpu is not preferred driver: tesla

real    0m0.005s
user    0m0.005s
sys     0m0.000s
```
I can be more precise in the measurement if needed, but it looks to be that `ghostdog` is very fast at providing this decision.

bcressey commented on packages/kmod-6.1-nvidia/nvidia-open-gpu-load-config.toml.in:
```diff
@@ -0,0 +1,16 @@
+objects-source = "__NVIDIA_MODULES__"
+load-only = true
+
+[nvidia-open-gpu.object-files."dummy.o"]
+link-objects = []
```
Why do we need a `dummy.o`?
```suggestion
```

yeazelm replied:
This was to avoid the empty object-files failing to parse. It is an artifact that I didn't adjust the configuration format to account for a configuration that doesn't really link. I'll shift the format around to be more flexible in light of the other comments in the review.

bcressey commented on packages/kmod-6.1-nvidia/open-gpu-depmod.conf.in:
```diff
@@ -0,0 +1,2 @@
+# Force the proprietary driver to be excluded since the open-gpu path will be preferred
```
I'm not sure why these `depmod` files are required if only one set of drivers is ever copied into the modules tree, as seems to be the goal of the current implementation.

yeazelm replied:
This was an insurance policy against accidental conflicting drivers but I think I've convinced myself now that we could back this out and trust the rest of the implementation to only provide one driver.

bcressey commented on sources/ghostdog/src/main.rs:
```diff
@@ -63,6 +141,44 @@ fn run() -> Result<()> {
+                    path: marker_path.clone(),
+                })?;
+
+                // Copy the driverdog configuration from the factory into /etc/drivers so the preferred driver is loaded
+                if !Path::new(DEFAULT_DRIVER_CONFIG_PATH).exists() {
+                    fs::create_dir_all(DEFAULT_DRIVER_CONFIG_PATH).context(
+                        error::CreateDirSnafu {
+                            path: DEFAULT_DRIVER_CONFIG_PATH,
+                        },
+                    )?;
+                }
+                let source_driver_path =
+                    Path::new(DRIVER_FACTORY_DIR).join(format!("{}.toml", driver_choice.clone()));
+                let etc_driver_path = Path::new(DEFAULT_DRIVER_CONFIG_PATH)
+                    .join(format!("{}.toml", driver_choice.clone()));
+                fs::copy(source_driver_path.clone(), etc_driver_path.clone()).context(
+                    error::CopyFileSnafu {
+                        s: source_driver_path,
+                        d: etc_driver_path,
+                    },
+                )?;
```
This is my major open concern with the change. There's no guarantee that the config file is in place before `driverdog` runs.

You are also making some assumptions (final `/etc` mounted, `/run` mounted) that are true for Bottlerocket but wouldn't hold for many other paths. In general I would avoid these kinds of filesystem operations in a udev context.

yeazelm replied:
I agree that these are not ideal in a udev context. I had a lot of difficulty getting systemd to enforce the strict ordering needed here and this was a way to make it work but it does do work we really shouldn't be doing via udev.

yeazelm replied:
I've moved away from `udev` so this should avoid these problems.

yeazelm reviewed (commented):
I've added comments here but also started breaking out the changes into smaller PRs since this one is a lot to review. See https://github.com/bottlerocket-os/bottlerocket-core-kit/pull/118 for the NVIDIA kmod packaging and https://github.com/bottlerocket-os/bottlerocket-core-kit/pull/119 for `driverdog` changes. Once those get merged, I'll rebase this PR on top of those and make this significantly smaller to review.

Commit:
ghostdog: add match-nvidia-driver argument

ghostdog can now look up PCI devices and confirm which NVIDIA driver
should be used. The match-nvidia-driver takes one argument, which type
of driver such as tesla or open-gpu and exits 0 if that matches the PCI
devices currently present. If those devices do not match the provided
driver, it exits 1. This can be used to set ExecCondition for things
like linking or loading drivers.

Signed-off-by: Matthew Yeazel <yeazelm@amazon.com>

Commit:
packages/kmod-6.1-nvidia: provide configuration for open-gpu

This adds upon the present logic to build the open-gpu driver to provide
configuration that driverdog can use to work with the open-gpu drivers.

Signed-off-by: Matthew Yeazel <yeazelm@amazon.com>

Commit:
packages/kmod-5.15-nvidia: provide configuration for open-gpu

This adds upon the present logic to build the open-gpu driver to provide
configuration that driverdog can use to work with the open-gpu drivers.

Signed-off-by: Matthew Yeazel <yeazelm@amazon.com>

Commit:
os: move NVIDIA specific services to their kmod packages

The os package should not be concerned with specifics to the NVIDIA kmod
module preferences. os will provide driverdog, but the configuration
files it reads will be provided by the package providing the modules to
keep that specific logic local to their domain.

Signed-off-by: Matthew Yeazel <yeazelm@amazon.com>

yeazelm commented:
^ Just pushed a bunch of changes, this should "reset" the PR to be the new approach and should remove the code already reviewed.

bcressey commented on packages/os/load-tesla-kernel-modules.service.in:
```diff
@@ -1,16 +1,17 @@
-After=link-kernel-modules.service
-Requires=link-kernel-modules.service
+After=link-tesla-kernel-modules.service
+Requires=link-tesla-kernel-modules.service
 # Disable manual restarts to prevent loading kernel modules
 # that weren't linked by the running system
 RefuseManualStart=true
 RefuseManualStop=true
 
 [Service]
 Type=oneshot
-ExecStart=/usr/bin/driverdog load-modules
+ExecCondition=/usr/bin/ghostdog match-nvidia-driver tesla
+ExecStart=/usr/bin/driverdog --modules-set nvidia-tesla load-modules
```
Now that these units are so specific to NVIDIA, they don't really fit with the general purpose code in `os`.

I would either move them to the `kmod-*-nvidia` packages, or else turn these into unit templates and adjust the logic a bit so we can do something like:
```
[Unit]
After=link-kernel-modules@%i.service
Requires=link-kernel-modules@%i.service

...

[Service]
Type=oneshot
ExecCondition=/usr/bin/ghostdog match-driver %I
ExecStart=/usr/bin/driverdog --modules-set %I load-modules
```

Then the `kmod-*-nvidia` packages could just add symlinks to instantiate the units they supported.

yeazelm replied:
This makes sense, I'll move them into the kmod packages. I had issues getting templates to block boot so we probably will opt for explicit files here to keep things straightforward even if its a bit of copying the same things over.

bcressey commented on sources/ghostdog/src/main.rs:
```diff
@@ -48,6 +53,40 @@ struct EbsDeviceNameArgs {
+}
+
+#[derive(PartialEq, Debug, Deserialize)]
+/// The GPU Device Data contains various features of the device. Only Name, Device ID, and Features are required
+/// a particular device
```
nit: missing word, guessing ...?
```suggestion
/// The GPU Device Data contains various features of the device. Only Name, Device ID, and Features are required
/// for a particular device
```

bcressey commented on sources/ghostdog/src/main.rs:
```diff
@@ -125,6 +168,46 @@ fn parse_device_name(device_info: &[u8], path: String) -> Result<String> {
+    let present_devices =
+        pciclient::list_devices(list_input).context(error::ListPciDevicesSnafu {})?;
+
+    for input_device in present_devices.iter() {
+        match open_gpu_devices {
+            SupportedDevicesConfiguration::OpenGpu(ref device_list) => {
+                let formatted_device_id = format!("0x{}", input_device.device());
+                for supported_device in device_list.iter() {
+                    if supported_device.device_id == formatted_device_id {
+                        return Ok("open-gpu".to_string());
+                    }
+                }
+            }
+        }
+    }
```
In a system with a large number of GPUs that weren't supported by the open driver, and a JSON file with a large number of different devices, we could end up doing a lot of extra work.

It would be more efficient to transform the open GPU device data into a hash set of device IDs. Then we would guarantee one pass through the list of supported devices, and one lookup per device.

yeazelm replied:
I tried putting these into `HashSet`s to ensure we didn't have negative performance even in cases where there are 16 devices and the runs all were taking between 5-6ms both as the Vec implementation as it stands vs the HashSet `contains()`. It stands to reason that HashSet's would be better so I can convert, but I don't think we gain a lot here with the ~700 devices listed in the current file, even in worst case.
```
    // Bogus PCI devices but "miss" to ensure there will be no match
    let present_devices: Vec<ListDevicesOutput> = vec![ListDevicesOutput {
                pci_slot: "00:00.0".to_string(),
                class: "0600".to_string(),
                vendor: "8086".to_string(),
                device: "1234".to_string(),
                program_interface: Some("00".to_string()),
                subsystem_vendor: Some("1d0f".to_string()),
                subsystem_device: Some("1237".to_string()),
                ..Default::default()
            };16];

    // reduce down to unique ids so we only iterate on uniqueness (down from 16 to 1 here)
    let unique_ids: HashSet<String> = present_devices.iter().map(|x| format!("0x{}", x.device()).clone()).collect();
    match open_gpu_devices {
            SupportedDevicesConfiguration::OpenGpu(ref device_list) => {
                let open_devices_list: HashSet<String> = device_list.iter().map(|x| x.device_id.clone()).collect();
                for input_device in unique_ids.iter() {
                    if open_devices_list.contains(input_device) {
                        return Ok("open-gpu".to_string());
                    }
            }
       }
    }
```
The compiler might be tricking me here and optimizing as well  (I did try creating the Vec of unique objects and still saw the same times for input devices) so I can test a bit more, but even so, 5ms runs for this when parsing the entire file is good considering the kernel module loads will take seconds. We might save more by saving the decision to a file to be checked before rerunning the deserialization and searching since this will be called several times by systemd.

Commit:
kmod-5.10-nvidia: relocate kernel module services

The os package doesn't need to concern itself with NVIDIA specific
loading behavior. It will provide driverdog, but the configurations read
by driverdog will be included with the specific kernel modules package
that provides the drivers described in the configuration.

Signed-off-by: Matthew Yeazel <yeazelm@amazon.com>

Commit:
kmod-5.15-nvidia: move kernel module services into package

The os package doesn't need to concern itself with NVIDIA specific
loading behavior. It will provide driverdog, but the configurations read
by driverdog will be included with the specific kernel modules package
that provides the drivers described in the configuration. This moves the
tesla and open-gpu services into the kmod-5.15-nvidia package instead.

Signed-off-by: Matthew Yeazel <yeazelm@amazon.com>

Commit:
kmod-6.1-nvidia: move kernel module services into package

The os package doesn't need to concern itself with NVIDIA specific
loading behavior. It will provide driverdog, but the configurations read
by driverdog will be included with the specific kernel modules package
that provides the drivers described in the configuration. This moves the
tesla and open-gpu services into the kmod-6.1-nvidia package instead.

Signed-off-by: Matthew Yeazel <yeazelm@amazon.com>

ytsssun commented on sources/ghostdog/src/main.rs:
```diff
@@ -1,12 +1,14 @@
 /*!
 ghostdog is a tool to manage ephemeral disks.
 It can be called as a udev helper program to identify ephemeral disks.
+It can also check if devices on the PCI bus match a particular NVIDIA driver
```
nit: closing "."
```suggestion
It can also check if devices on the PCI bus match a particular NVIDIA driver.
```

ytsssun commented on sources/ghostdog/src/main.rs:
```diff
@@ -125,6 +168,51 @@ fn parse_device_name(device_info: &[u8], path: String) -> Result<String> {
+        .iter()
+        .map(|x| format!("0x{}", x.device().to_uppercase()).clone());
+
+    let (device_list, driver_choice) = match &open_gpu_devices {
```
nit: should we name it device_set instead indicating it is a set not a list? The `device_list.contains(&input_device)` felt expensive by reading the name, but it in fact yields O(1) operation for the look up.

yeazelm replied:
Corrected.

yeazelm commented:
^ updated to reflect comments. This now holds back open gpu use in some cases where it is known to cause issues. Also moved the kernel module services to the kmod packages.

ytsssun commented on sources/ghostdog/src/main.rs:
```diff
@@ -125,6 +168,51 @@ fn parse_device_name(device_info: &[u8], path: String) -> Result<String> {
+        .vendor(NVIDIA_VENDOR_ID)
+        .build();
+    let present_devices =
+        pciclient::list_devices(list_input).context(error::ListPciDevicesSnafu {})?;
```
nit: for variant that only has `source` as member field, you can do 
```suggestion
        pciclient::list_devices(list_input).context(error::ListPciDevicesSnafu)?;
```

Same for L181

bcressey commented on packages/kmod-5.15-nvidia/link-tesla-kernel-modules.service.in:
```diff
@@ -0,0 +1,19 @@
+[Unit]
+Description=Link additional kernel modules
```
Ideally the descriptions for all these units would reflect that they're linking and loading NVIDIA tesla vs. open-gpu drivers. Otherwise the logs and console output will be unclear.

yeazelm replied:
Makes sense! I'll update these!

bcressey commented on packages/kmod-5.15-nvidia/kmod-5.15-nvidia.spec:
```diff
@@ -300,6 +342,13 @@ popd
+%{_cross_unitdir}/copy-open-gpu-kernel-modules.service
+%{_cross_unitdir}/load-open-gpu-kernel-modules.service
+
+
```
nit: so the spec files are back to matching except for kernel versions
```suggestion
```

yeazelm replied:
Nice catch! Will remove.

bcressey commented on sources/ghostdog/src/main.rs:
```diff
@@ -125,6 +168,51 @@ fn parse_device_name(device_info: &[u8], path: String) -> Result<String> {
+    let present_devices =
+        pciclient::list_devices(list_input).context(error::ListPciDevicesSnafu {})?;
+
+    // If there a multiple devices with the same ID, dedup them to minimize iterations
+    let mut unique_ids = present_devices
+        .iter()
+        .map(|x| format!("0x{}", x.device().to_uppercase()).clone());
```
Does this really dedupe IDs like the comment says?

yeazelm replied:
I had the `collect::<HashSet<_>>()` at one point and was troubleshooting this issue I found with `to_uppercase()` and forgot to put it back in. I'll update, thanks for the catch!

bcressey commented on sources/ghostdog/src/main.rs:
```diff
@@ -125,6 +168,51 @@ fn parse_device_name(device_info: &[u8], path: String) -> Result<String> {
+    };
+
+    if unique_ids.any(|input_device| device_list.contains(&input_device)) {
+        Ok(driver_choice.to_string())
```
nit: I think it'd be more clear to use the string here and not bother with the tuple:
```suggestion
        Ok("open-gpu".to_string())
```

yeazelm replied:
I spoke with @cbgbt offline about this approach. The rationale here was to make this ready to take other driver choices if needed in the future since we are using a single variant of an enum. Nonetheless, we don't have other variants yet so we can simplify the code as it is today.

yeazelm commented:
^ Pushed a new set of commits, this renames the systemd services to be more clear and fixes the unique id bug. Validated on g5.2xlarge which used proprietary and a g4d.24xlarge that chose the open GPU drivers.

yeazelm commented:
^ rebase on develop

bcressey commented on sources/ghostdog/src/main.rs:
```diff
@@ -143,6 +186,50 @@ fn parse_device_name(device_info: &[u8], path: String) -> Result<String> {
+    let present_devices =
+        pciclient::list_devices(list_input).context(error::ListPciDevicesSnafu)?;
+
+    // If there a multiple devices with the same ID, dedup them to minimize iterations
```
nit:
```suggestion
    // If there are multiple devices with the same ID, dedup them to minimize iterations
```
